File: 1095 find-in-mountain-array/solution.java
/**
 * // This is MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface MountainArray {
 *     public int get(int index) {}
 *     public int length() {}
 * }
 */

class Solution {
    public int findInMountainArray(int target, MountainArray m) {
        int low=0,high=m.length()-1;
        while(low<high) {
            int mid = low+(high-low)/2;
            if(m.get(mid)<m.get(mid+1)) low=mid+1;
            else high=mid;
        }
        int peak = low;
        int ans = bin(m,target,0,peak,true);
        if(ans!=-1) return ans;
        return bin(m,target,peak+1,m.length()-1,false);
    }

    private int bin(MountainArray m, int target,int low,int high,boolean asc)
    {
        while(low<=high) {
            int mid = low+(high-low)/2;
            int ans =m.get(mid);
            if(ans==target) return mid;
            if(asc) {
                if(ans<target) {
                    low=mid+1;
                } else {
                    high=mid-1;
                }
            } else {
                if(ans<target) {
                    high=mid-1;
                } else {
                    low=mid+1;
                }
            }
        }
        return -1;
    }
}

File: 1095 find-in-mountain-array/solution.rs
/**
 * // This is the MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 *  struct MountainArray;
 *  impl MountainArray {
 *     fn get(index:i32)->i32;
 *     fn length()->i32;
 * };
 */

impl Solution {
    pub fn find_in_mountain_array(target: i32, mo: &MountainArray) -> i32 {
        let mut low = 0;
        let mut high =  mo.length() - 1;
        while low<high {
            let mid = low+(high-low)/2;
            if mo.get(mid) < mo.get(mid + 1) {
                low = mid+1;
            } else {
                high = mid;
            }
        }
        let peak = low;
        let ans = Self::bin(mo, target, 0, peak, true);
        if ans != -1 {
            return ans;
        }
       Self::bin(mo, target, peak + 1, mo.length() - 1, false)
    }
    fn bin(mo: &MountainArray, target: i32, mut low: i32, mut high: i32, asc: bool) -> i32 {
        while low <= high {
            let mid = low+(high-low)/2;
            let val = mo.get(mid);
            if val == target {
                return mid;
            }
            if asc {
                if val < target {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            } else {
                if val < target {
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            }
        }
        -1
    }
}
